### ISSUE TITLE
PG::InternalError: ERROR:  unhandled cypher(cstring) function call

### ISSUE THREAD

**Issue by calebl** [marpori/rails_age#2]
Every time I run "bin/rails console" it seems to lose the age extension being available for the database. 

If I run the following two commands, everything works again properly
```
ActiveRecord::Base.connection.execute("LOAD 'age';")
```

```
ActiveRecord::Base.connection.execute("SET search_path = ag_catalog, age_schema, public;")
```

I'm seeing the same thing when I try to run a rake task that is creating nodes. Am I missing something that should be in an initializer? 

This is at the top of my schema file:
<img width="598" alt="image" src="https://github.com/user-attachments/assets/fe9f9063-50c8-4f85-bf3c-eed3998ed01d">

(Currently on v0.6.4 of the gem)
Rails 7.1.4
Ruby 3.3.5
Postgres 16.3

**Comment by btihen** [comment]
True, I'll mark that as a KNOWN issue.  And document the solution.
I am not sure why `bin/rails c` doesn't load all the DB settings, but `bin/rails s` loads all the DB settings.  

I am guessing we need to do something to load the DB settings (or just change rails c to load both:
```
ActiveRecord::Base.connection.execute("LOAD 'age';")
ActiveRecord::Base.connection.execute("SET search_path = ag_catalog, age_schema, public;")
```

I will look into this.

**Comment by calebl** [comment]
This is what I've got for my database.yml file:
<img width="607" alt="image" src="https://github.com/user-attachments/assets/35730051-0028-415b-8c16-6030f815d1ae">

I'm using `asdf` to load ruby and postgres. I'll try updating to Rails 7.2.2 and see if that gives any different results.

**Comment by calebl** [comment]
same result after upgrading to Rails 7.2.2

**Comment by calebl** [comment]
Tried deleting my bootsnap cache to see if that would help, but no change with that.

**Comment by calebl** [comment]
Based on this issue: https://github.com/apache/age/issues/41 It looks like `load 'age';` needs to be called for each fresh connection to the database. 

If I add `ActiveRecord::Base.connection.execute("LOAD 'age';")` at the end of my `config/environment.rb` file it fixes the issue for me. Not sure if that is the best place for it or not.

**Comment by btihen** [comment]
Cool - I'll check into that.  It sounds reasonable.  But I don't understand the inconsistent behavior. 
I will try to see if I can understand what might be different between our setups.

**Comment by btihen** [comment]
Are you using a compiled version of AGE or the Docker file (I am using the Mac Version of the Docker PostgresAGE?  
Are you using Windows or Mac OS or Linux?

It seems you are right -- it seems to be related to the DB.  
For example, when I use `bin/rails db` I need to first run:
 `SET search_path = ag_catalog, "$user", public;`
but not 
 `LOAD 'age';`

Here is the log:
```
bin/rails db

Password for user postgresUser: 

psql (14.13 (Homebrew), server 16.3 (Debian 16.3-1.pgdg120+1))

Type "help" for help.

--- first query
# SELECT * FROM cypher('age_schema', $$ MATCH (find:Person) RETURN find $$) AS (find agtype);
ERROR:  function cypher(unknown, unknown) does not exist
LINE 1: SELECT * FROM cypher('age_schema', $$ MATCH (find:Person) RE...
                      ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.

--- set paths
# SET search_path = ag_catalog, "$user", public;
SET

--- second select
# SELECT * FROM cypher('age_schema', $$ MATCH (find:Person) RETURN find $$) AS (find agtype);
--- get expected results
```

I will experiment some alternatives, but for now I think your workaround looks very promising.

**Comment by calebl** [comment]
I am on Mac using the compiled version of AGE.

For your example, I'm curious what happens if you do not run `SET search_path` in between the first and second query. From what I was seeing, the first query will fail and then the second will pass. Running `Load 'age';` prior to the first query is the only thing I've found that will prevent an error happening in the first query.
